app/context/authContext.tsx

```diff
@@ -0,0 +1,175 @@
+  login: ({
+    // eslint-disable-next-line no-unused-vars
+    email, // eslint-disable-next-line no-unused-vars
+    password, // eslint-disable-next-line no-unused-vars
+    role,
+  }: {
+    email: string;
+    password: string;
+    role: Role;
+  }) => void;
+  getUserStatus: QueryStatus;
+  getUserError: Error | null;
+  logout: () => void;
+  loginStatus: MutationStatus;
+  loginError: Error | null;
+  logoutStatus: MutationStatus;
+  logoutError: Error | null;
+}
+
+export const AuthContext = createContext<AuthContextType | undefined>(undefined);
+
+function getUrl(role: Role): string {
+  let url = baseUrl;
+  if (role === 'member') url = `${baseUrl}`;
+  else if (role === 'content_manager') url = `${baseUrl}/content`;
+  else if (role === 'admin') url = `${baseUrl}/admin`;
+  return url;
+}
+
+export const AuthProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
+  const getUser = async (): Promise<UserType | null | undefined> => {
+    const session = await getSession();
+    if (!session?.token || !session.role) throw new Error('No session found');
+    const response = await fetch(`${getUrl(session.role)}/user`, {
+      headers: {
+        Accept: 'application/json',
+        Referer: 'learnhub.mk',
+        Authorization: `Bearer ${session.token}`,
+      },
+      // credentials: 'include',
+    });
+    if (!response.ok) {
+      throw new Error('An error occurred while fetching user data');
+    }
+    const data = await response.json();
+    return data.data.user;
+  };
+
+  const {
+    data: user,
+    status: getUserStatus,
+    error: getUserError,
+    refetch: refetchUser,
+  } = useQuery({
+    queryKey: ['user'],
+    queryFn: getUser,
+  });
+
+  const {
+    mutate: login,
+    error: loginError,
+    status: loginStatus,
+  } = useMutation({
+    mutationKey: ['login'],
+    mutationFn: async ({
+      email,
+      password,
+      role,
+    }: {
+      email: string;
+      password: string;
+      role: Role;
+    }) => {
+      return fetch(`${getUrl(role)}/login`, {
+        method: 'POST',
+        headers: {
+          Accept: 'application/json',
+          Referer: 'learnhub.mk',
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({ email, password }),
+      });
+    },
+    onSuccess: (data) =>
+      data.json().then(async (res) => {
+        const token = res.access_token;
+        if (token) await setSession({ token, role: res.data.user.role });
+        await refetchUser();
+      }),
+    onError: (error) => {
+      throw new Error(error?.message || 'An error occurred while logging in');
+    },
+  });
+
+  const {
+    status: logoutStatus,
+    mutate: logout,
+    error: logoutError,
+  } = useMutation({
+    mutationKey: ['logout'],
+    mutationFn: async () => {
+      const session = await getSession();
+      if (!session?.token || !session.role) throw new Error('No session found');
+      await fetch(`${getUrl(session.role)}/logout`, {
```

Comment:
we can create a middleware that would check the session before  each request

Reply:
@edichoska Shall we do something like here: https://github.com/laravel/breeze-next/blob/master/src/hooks/auth.js

In this example in the `useAuth` hook we pass `middleware` prop that decide if the user needs to be authenticated, and if so we check if the `getUserStatus` from the `AuthContext` is `success` and if so we don't do anything, otherwise we redirect the user to the login route. 
This can create protected routes, everywhere we need the user to be logged in, we can call `useAuth` with passed `middleware: auth` as prop.

Example:

```
export const useAuth = ({
  middleware,
  redirectIfAuthenticatedTo,
}: {
  middleware: 'auth' | 'guest';
  redirectIfAuthenticatedTo: string;
}) => {
  const context = useContext(AuthContext);
  const router = useRouter();
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  const { getUserStatus } = context;

  useEffect(() => {
    if (middleware !== 'auth') return;
    if (getUserStatus === 'pending') return;
    if (getUserStatus === 'error') {
      router.push('/login');
    }
    if (getUserStatus === 'success' && redirectIfAuthenticatedTo)
      router.push(redirectIfAuthenticatedTo);
  }, [getUserStatus, middleware, redirectIfAuthenticatedTo, router]);
  return { ...context };
};

```

This is all in the client-side and don't have anything to do with the nextjs middleware. But in the nextjs middleware, we can check to see the expiration date of the session, and refresh it if it is expiring in a certain time?

Reply:
@stefanatanasovmk, from my side, we can go with the proposed solution, and integrating server-side checks with Next.js middleware for session expiration and refresh could provide a more robust solution.